### DISCUSSION TITLE
Error: Cannot load Firebird client library: 'fbclient.dll'.

### DISCUSSION THREAD

**Discussion by faroldesenvolvimento5** [asfernandes/node-firebird-drivers#113] [Q&A]
With node-firebird-driver-native installed, i tried to run the [test](https://github.com/asfernandes/node-firebird-drivers/blob/master/packages/node-firebird-driver-native/README.md) but this happened:

(node:2376) UnhandledPromiseRejectionWarning: Error: Cannot load Firebird client library: 'fbclient.dll'.
    at new ClientImpl (C:\Users\user\Documents\Projects\Firebird4\node_modules\node-firebird-driver-native\dist\lib\client.js:28:26)
    at createNativeClient (C:\Users\user\Documents\Projects\Firebird4\node_modules\node-firebird-driver-native\dist\lib\client.js:21:12)
    at test (file:///C:/Users/user/Documents/Projects/Firebird4/index.mjs:10:17)
    at file:///C:/Users/user/Documents/Projects/Firebird4/index.mjs:38:1
    at ModuleJob.run (internal/modules/esm/module_job.js:169:25)
    at async Loader.import (internal/modules/esm/loader.js:177:24)
    at async Object.loadESM (internal/process/esm_loader.js:68:5)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:2376) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, 
use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:2376) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

![image](https://user-images.githubusercontent.com/85496332/137480094-e3b37766-4726-4bda-b042-839aa48f3998.png)

![image](https://user-images.githubusercontent.com/85496332/137480863-b698e212-1816-49a2-9cfc-7aabc95f440a.png)

Yes, it is a Firebird4 fbclient.dll, thanks for the attention.

**Comment by asfernandes** [comment]
It must be on the PATH (or in the same directory of the `.exe` file).

Or, use the absolute path instead of `getDefaultLibraryFilename()`.

**Reply by faroldesenvolvimento5** [reply]
> It must be on the PATH (or in the same directory of the `.exe` file).
> 
> Or, use the absolute path instead of `getDefaultLibraryFilename()`.

Using FS to read the file with "getDefaultLibraryFilename()" works correctly, already  tried with absolute path too, but it still doesn't work.

**Reply by asfernandes** [reply]
> Using FS to read the file with "getDefaultLibraryFilename()" works correctly, already tried with absolute path too, but it still doesn't work.

I do not understand what you mean. What worked and what not worked?

**Reply by faroldesenvolvimento5** [reply]
> > Using FS to read the file with "getDefaultLibraryFilename()" works correctly, already tried with absolute path too, but it still doesn't work.
> 
> I do not understand what you mean. What worked and what not worked?

I tried "fs.statSync()" with "getDefaultLibraryFilename()" and absolute path, both returned the dll file info, but neither worked with "createNativeClient()".

**Reply by asfernandes** [reply]
And what? As I said, to load a DLL with not-absolute path, it must be in same directory of `.exe` file or its directory must be in `PATH` environment variable.

**Reply by faroldesenvolvimento5** [reply]
> And what? As I said, to load a DLL with not-absolute path, it must be in same directory of `.exe` file or its directory must be in `PATH` environment variable.

But even with absolute path it still doesn't work, 

`(node:11256) UnhandledPromiseRejectionWarning: Error: Cannot load Firebird client library: 'C:\Users\user\Documents\Projects\Firebird4\fbclient.dll'.`

**Comment by faroldesenvolvimento5** [comment]
Still trying but the error continues...
Code:
![image](https://user-images.githubusercontent.com/85496332/137723824-a52cdff3-7bff-4b07-8a6c-f4cdbab10efd.png)
Output:
`File Path: C:\Users\user\Documents\Projects\Firebird4\fbclient.dll
File Info:
Stats {
  dev: 2487996224,
  mode: 33206,
  nlink: 1,
  uid: 0,
  gid: 0,
  rdev: 0,
  blksize: 4096,
  ino: 2533274791030627,
  size: 1592320,
  blocks: 3112,
  atimeMs: 1634556890872.223,
  mtimeMs: 1622444400000,
  ctimeMs: 1634304466307.2659,
  birthtimeMs: 1634221204336.0437,
  atime: 2021-10-18T11:34:50.872Z,
  mtime: 2021-05-31T07:00:00.000Z,
  ctime: 2021-10-15T13:27:46.307Z,
  birthtime: 2021-10-14T14:20:04.336Z
}
(node:13924) UnhandledPromiseRejectionWarning: Error: Cannot load Firebird client library: 'C:\Users\user\Documents\Projects\Firebird4\fbclient.dll'.
    at new ClientImpl (C:\Users\user\Documents\Projects\Firebird4\node_modules\node-firebird-driver-native\dist\lib\client.js:28:26)
    at Object.createNativeClient (C:\Users\user\Documents\Projects\Firebird4\node_modules\node-firebird-driver-native\dist\lib\client.js:21:12)   
    at test (C:\Users\user\Documents\Projects\Firebird4\index.js:14:26)
    at Object.<anonymous> (C:\Users\user\Documents\Projects\Firebird4\index.js:42:1)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)
    at internal/main/run_main_module.js:17:47
(Use `node --trace-warnings ...` to show where the warning was created)
(node:13924) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:13924) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.`

As the output shows, the DLL is there, with an "absolute" path, the path is correct, FS reads the DLL file info, but the lib can't load the client. It's not a problem of path.

**Reply by asfernandes** [reply]
Are you sure you are using node and Firebird of the same architecture (32 vs 64 bits)?
